core/trino-main/src/main/java/io/trino/execution/buffer/SpoolingExchangeOutputBuffer.java

```diff
@@ -166,34 +167,51 @@ public void enqueue(int partition, List<Slice> pages)
     @Override
     public void setNoMorePages()
     {
-        stateMachine.noMorePages();
-        destroy();
+        if (stateMachine.noMorePages()) {
+            exchangeSink.finish().whenComplete((value, failure) -> {
+                try {
+                    if (failure != null) {
+                        stateMachine.fail(failure);
+                    }
+                    else {
+                        stateMachine.finish();
+                    }
+                }
+                finally {
+                    updateMemoryUsage(0);
+                }
+            });
+        }
     }
 
     @Override
     public void destroy()
     {
-        if (stateMachine.finish()) {
-            try {
-                exchangeSink.finish();
-            }
-            finally {
-                updateMemoryUsage(exchangeSink.getMemoryUsage());
-            }
+        if (stateMachine.getState().canAddPages()) {
+            // This situation is possible if the task has been cancelled
+            // Task cancellation is not supported (and not expected to be requested by the scheduler)
+            // when external exchange is active as the task output is expected to be deterministic
+            // As a safety precaution abort the sink to mark the output as invalid
+            abort();
         }
     }
 
     @Override
     public void abort()
     {
-        if (stateMachine.abort()) {
+        exchangeSink.abort().whenComplete((value, failure) -> {
```

Comment:
can there be a race between `whenComplete` for `abort` and `finish`?  IMO it can if implementor decides (reasonably) to use executor for notifications. I don't think you have any ordering guarantees. For example, `finish` completed and notification is about to execute, but then `abort` was called and it's notification was received first.

 Do we miss some extra states as `ABORTING` of `FINISHING` to prevent races?

Reply:
The `abort` method is only expected to be called when the task got aborted or failed. The `destroy` method is expected to be called only when the task got cancelled. It doesn't seem like this makes a huge difference in what state the buffer will end up since the task has already finished execution.